In [145]:
import pandas as pd


In [146]:
df = pd.read_csv("preprocessed_data.csv")
df.shape


(6268, 10)

In [147]:
df.head()


,CGPA,Internship_Experience,Projects,Certifications,AptitudeTestScore,Communication_Skills,Placement_Status,Technical_Skills,Placement_Label,Internship_Flag
0,7.7,1,1,0,69,4.0,NotPlaced,2.592593,0,1
1,7.6,2,3,1,80,4.3,NotPlaced,7.777778,0,1
2,7.0,1,1,1,64,4.1,NotPlaced,3.703704,0,1
3,7.9,1,3,2,90,4.8,Placed,7.777778,1,1
4,6.7,0,2,1,71,3.7,NotPlaced,4.074074,0,0


In [148]:
df.columns


Index(['CGPA', 'Internship_Experience', 'Projects', 'Certifications',
       'AptitudeTestScore', 'Communication_Skills', 'Placement_Status',
       'Technical_Skills', 'Placement_Label', 'Internship_Flag'],
      dtype='object')

In [149]:
X = df.drop(columns=["Placement_Label", "Placement_Status"])
y = df["Placement_Label"]
X.shape, y.shape


((6268, 8), (6268,))

In [150]:
X.columns


Index(['CGPA', 'Internship_Experience', 'Projects', 'Certifications',
       'AptitudeTestScore', 'Communication_Skills', 'Technical_Skills',
       'Internship_Flag'],
      dtype='object')

In [151]:
from sklearn.model_selection import train_test_split


In [152]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)


In [153]:
X_train.shape, X_test.shape


((5014, 8), (1254, 8))

In [154]:
y_train.shape, y_test.shape


((5014,), (1254,))

In [155]:
y_train.value_counts(normalize=True)
y_test.value_counts(normalize=True)


Placement_Label
0    0.641148
1    0.358852
Name: proportion, dtype: float64

# LOGISTIC REGRESSION

In [156]:
from sklearn.linear_model import LogisticRegression
log_reg = LogisticRegression(
    max_iter=1000,
    random_state=42
)


In [157]:
log_reg.fit(X_train, y_train)


LogisticRegression(max_iter=1000, random_state=42)

In [158]:
y_pred = log_reg.predict(X_test)
y_prob = log_reg.predict_proba(X_test)
placement_prob = y_prob[:, 1]


In [159]:
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    confusion_matrix,
    classification_report
)


In [160]:
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

accuracy, precision, recall, f1


(0.7759170653907496, 0.7172236503856041, 0.62, 0.66507747318236)

In [161]:
conf_matrix = confusion_matrix(y_test, y_pred)
conf_matrix


array([[694, 110],
       [171, 279]], dtype=int64)

In [162]:
print(classification_report(y_test, y_pred))


              precision    recall  f1-score   support

           0       0.80      0.86      0.83       804
           1       0.72      0.62      0.67       450

    accuracy                           0.78      1254
   macro avg       0.76      0.74      0.75      1254
weighted avg       0.77      0.78      0.77      1254



# DECISION TREE

In [163]:
from sklearn.tree import DecisionTreeClassifier


In [164]:
dt_model = DecisionTreeClassifier(
    max_depth=5,
    min_samples_split=20,
    random_state=42
)


In [165]:
dt_model.fit(X_train, y_train)


DecisionTreeClassifier(max_depth=5, min_samples_split=20, random_state=42)

In [166]:
y_pred_dt = dt_model.predict(X_test)
y_prob_dt = dt_model.predict_proba(X_test)
placement_prob_dt = y_prob_dt[:, 1]


In [167]:
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    confusion_matrix,
    classification_report
)


In [168]:
accuracy_dt = accuracy_score(y_test, y_pred_dt)
precision_dt = precision_score(y_test, y_pred_dt)
recall_dt = recall_score(y_test, y_pred_dt)
f1_dt = f1_score(y_test, y_pred_dt)

accuracy_dt, precision_dt, recall_dt, f1_dt


(0.7671451355661882,
 0.7146739130434783,
 0.5844444444444444,
 0.6430317848410758)

In [169]:
conf_matrix_dt = confusion_matrix(y_test, y_pred_dt)
conf_matrix_dt


array([[699, 105],
       [187, 263]], dtype=int64)

In [170]:
print(classification_report(y_test, y_pred_dt))


              precision    recall  f1-score   support

           0       0.79      0.87      0.83       804
           1       0.71      0.58      0.64       450

    accuracy                           0.77      1254
   macro avg       0.75      0.73      0.74      1254
weighted avg       0.76      0.77      0.76      1254



# SVM

In [171]:
from sklearn.svm import SVC


In [172]:
svm_model = SVC(
    kernel="rbf",
    probability=True,
    C=1.0,
    gamma="scale",
    random_state=42
)


In [173]:
svm_model.fit(X_train, y_train)


SVC(probability=True, random_state=42)

In [174]:
y_pred_svm = svm_model.predict(X_test)
y_prob_svm = svm_model.predict_proba(X_test)
placement_prob_svm = y_prob_svm[:, 1]
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    confusion_matrix,
    classification_report
)

accuracy_svm = accuracy_score(y_test, y_pred_svm)
precision_svm = precision_score(y_test, y_pred_svm)
recall_svm = recall_score(y_test, y_pred_svm)
f1_svm = f1_score(y_test, y_pred_svm)

accuracy_svm, precision_svm, recall_svm, f1_svm


(0.7639553429027113,
 0.7115384615384616,
 0.5755555555555556,
 0.6363636363636364)

In [175]:
conf_matrix_svm = confusion_matrix(y_test, y_pred_svm)
conf_matrix_svm


array([[699, 105],
       [191, 259]], dtype=int64)

In [176]:
print(classification_report(y_test, y_pred_svm))


              precision    recall  f1-score   support

           0       0.79      0.87      0.83       804
           1       0.71      0.58      0.64       450

    accuracy                           0.76      1254
   macro avg       0.75      0.72      0.73      1254
weighted avg       0.76      0.76      0.76      1254



# KNN

In [177]:
from sklearn.neighbors import KNeighborsClassifier


In [178]:
knn_model = KNeighborsClassifier(
    n_neighbors=7,
    metric="minkowski",
    p=2
)


In [179]:
knn_model.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=7)

In [180]:
y_pred_knn = knn_model.predict(X_test)

In [181]:
y_prob_knn = knn_model.predict_proba(X_test)
placement_prob_knn = y_prob_knn[:, 1]


In [182]:
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    confusion_matrix,
    classification_report
)

accuracy_knn = accuracy_score(y_test, y_pred_knn)
precision_knn = precision_score(y_test, y_pred_knn)
recall_knn = recall_score(y_test, y_pred_knn)
f1_knn = f1_score(y_test, y_pred_knn)

accuracy_knn, precision_knn, recall_knn, f1_knn


(0.7360446570972887,
 0.6386946386946387,
 0.6088888888888889,
 0.6234357224118316)

In [183]:
conf_matrix_knn = confusion_matrix(y_test, y_pred_knn)
conf_matrix_knn


array([[649, 155],
       [176, 274]], dtype=int64)

In [184]:
print(classification_report(y_test, y_pred_knn))


              precision    recall  f1-score   support

           0       0.79      0.81      0.80       804
           1       0.64      0.61      0.62       450

    accuracy                           0.74      1254
   macro avg       0.71      0.71      0.71      1254
weighted avg       0.73      0.74      0.73      1254



# RANDOM FOREST

In [185]:
from sklearn.ensemble import RandomForestClassifier


In [186]:
rf_model = RandomForestClassifier(
    n_estimators=100,
    max_depth=8,
    min_samples_split=10,
    random_state=42,
    n_jobs=-1
)


In [187]:
rf_model.fit(X_train, y_train)



RandomForestClassifier(max_depth=8, min_samples_split=10, n_jobs=-1,
                       random_state=42)

In [188]:
y_pred_rf = rf_model.predict(X_test)


In [281]:
for t in [0.5, 0.45, 0.4, 0.35]:
    y_pred_new = (y_prob_rf > t).astype(int)
    
    print(f"\nThreshold: {t}")
    print("Recall:", recall_score(y_test, y_pred_new))
    print("Precision:", precision_score(y_test, y_pred_new))
    print("F1:", f1_score(y_test, y_pred_new))


Threshold: 0.5
Recall: 0.6177777777777778
Precision: 0.7220779220779221
F1: 0.665868263473054

Threshold: 0.45
Recall: 0.6533333333333333
Precision: 0.6774193548387096
F1: 0.665158371040724

Threshold: 0.4
Recall: 0.7133333333333334
Precision: 0.6381709741550696
F1: 0.6736621196222455

Threshold: 0.35
Recall: 0.7666666666666667
Precision: 0.6149732620320856
F1: 0.6824925816023739


In [282]:
y_prob_rf = rf_model.predict_proba(X_test)[:, 1]

THRESHOLD = 0.35 

y_pred_rf = (y_prob_rf > THRESHOLD).astype(int)
placement_prob_rf = y_prob_rf


In [283]:
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    confusion_matrix,
    classification_report
)

accuracy_rf = accuracy_score(y_test, y_pred_rf)
precision_rf = precision_score(y_test, y_pred_rf)
recall_rf = recall_score(y_test, y_pred_rf)
f1_rf = f1_score(y_test, y_pred_rf)

accuracy_rf, precision_rf, recall_rf, f1_rf


(0.7440191387559809,
 0.6149732620320856,
 0.7666666666666667,
 0.6824925816023739)

In [284]:
conf_matrix_rf = confusion_matrix(y_test, y_pred_rf)
conf_matrix_rf


array([[588, 216],
       [105, 345]], dtype=int64)

In [285]:
print(classification_report(y_test, y_pred_rf))


              precision    recall  f1-score   support

           0       0.85      0.73      0.79       804
           1       0.61      0.77      0.68       450

    accuracy                           0.74      1254
   macro avg       0.73      0.75      0.73      1254
weighted avg       0.76      0.74      0.75      1254



# catboost

In [251]:
pip install catboost

Note: you may need to restart the kernel to use updated packages.


In [286]:
from catboost import CatBoostClassifier


In [287]:
cat_model = CatBoostClassifier(
    iterations=500,
    depth=6,
    learning_rate=0.05,
    random_state=42,
    verbose=0
)

In [288]:
cat_model.fit(X_train, y_train)


In [289]:
y_pred_cat = cat_model.predict(X_test)
y_prob_cat = cat_model.predict_proba(X_test)

placement_prob_cat = y_prob_cat[:, 1]


In [290]:
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    confusion_matrix,
    classification_report
)

accuracy_cat = accuracy_score(y_test, y_pred_cat)
precision_cat = precision_score(y_test, y_pred_cat)
recall_cat = recall_score(y_test, y_pred_cat)
f1_cat = f1_score(y_test, y_pred_cat)

accuracy_cat, precision_cat, recall_cat, f1_cat


(0.7687400318979266, 0.696078431372549, 0.6311111111111111, 0.662004662004662)

In [291]:
conf_matrix_cat = confusion_matrix(y_test, y_pred_cat)
conf_matrix_cat


array([[680, 124],
       [166, 284]], dtype=int64)

In [292]:
print(classification_report(y_test, y_pred_cat))


              precision    recall  f1-score   support

           0       0.80      0.85      0.82       804
           1       0.70      0.63      0.66       450

    accuracy                           0.77      1254
   macro avg       0.75      0.74      0.74      1254
weighted avg       0.77      0.77      0.77      1254



# Comparison

In [293]:
import pandas as pd
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

model_comparison = pd.DataFrame({
    "Model": [
        "Logistic Regression",
        "Decision Tree",
        "SVM",
        "KNN",
        "Random Forest",
        "CatBoost"
    ],
    "Accuracy": [
        accuracy_score(y_test, y_pred),
        accuracy_score(y_test, y_pred_dt),
        accuracy_score(y_test, y_pred_svm),
        accuracy_score(y_test, y_pred_knn),
        accuracy_score(y_test, y_pred_rf),
        accuracy_score(y_test, y_pred_cat)
    ],
    "Precision (Placed)": [
        precision_score(y_test, y_pred),
        precision_score(y_test, y_pred_dt),
        precision_score(y_test, y_pred_svm),
        precision_score(y_test, y_pred_knn),
        precision_score(y_test, y_pred_rf),
        precision_score(y_test, y_pred_cat)
    ],
    "Recall (Placed)": [
        recall_score(y_test, y_pred),
        recall_score(y_test, y_pred_dt),
        recall_score(y_test, y_pred_svm),
        recall_score(y_test, y_pred_knn),
        recall_score(y_test, y_pred_rf),
        recall_score(y_test, y_pred_cat)
    ],
    "F1-score (Placed)": [
        f1_score(y_test, y_pred),
        f1_score(y_test, y_pred_dt),
        f1_score(y_test, y_pred_svm),
        f1_score(y_test, y_pred_knn),
        f1_score(y_test, y_pred_rf),
        f1_score(y_test, y_pred_cat)
    ]
})

model_comparison.sort_values(by="F1-score (Placed)", ascending=False)


,Model,Accuracy,Precision (Placed),Recall (Placed),F1-score (Placed)
4,Random Forest,0.744019,0.614973,0.766667,0.682493
0,Logistic Regression,0.775917,0.717224,0.620000,0.665077
5,CatBoost,0.768740,0.696078,0.631111,0.662005
1,Decision Tree,0.767145,0.714674,0.584444,0.643032
2,SVM,0.763955,0.711538,0.575556,0.636364
3,KNN,0.736045,0.638695,0.608889,0.623436


In [294]:
print("CatBoost Train Accuracy:", cat_model.score(X_train, y_train))
print("CatBoost Test Accuracy:", cat_model.score(X_test, y_test))


CatBoost Train Accuracy: 0.8314718787395293
CatBoost Test Accuracy: 0.7687400318979266


In [295]:
from sklearn.model_selection import cross_val_score

cv_scores = cross_val_score(cat_model, X, y, cv=5)
print("Cross Validation Accuracy:", cv_scores.mean())


Cross Validation Accuracy: 0.7575005314199669


In [296]:
from sklearn.metrics import roc_auc_score

auc_cat = roc_auc_score(y_test, placement_prob_cat)
print("CatBoost ROC-AUC:", auc_cat)


CatBoost ROC-AUC: 0.8106826976229962


In [297]:
importance = cat_model.get_feature_importance()
feature_df = pd.DataFrame({
    "Feature": X.columns,
    "Importance": importance
}).sort_values(by="Importance", ascending=False)

feature_df


,Feature,Importance
4,AptitudeTestScore,27.421435
0,CGPA,20.746433
5,Communication_Skills,19.288644
6,Technical_Skills,11.887374
3,Certifications,7.884240
2,Projects,5.873384
1,Internship_Experience,4.233865
7,Internship_Flag,2.664624


The CatBoost model achieved a training accuracy of 83.15% and a testing accuracy of 76.87%. The gap of approximately 6.3% between training and testing accuracy maintains reasonable generalization.
The cross-validation accuracy of 75.75% is close to the test accuracy.The ROC-AUC score of 0.8107 shows good discriminative ability in distinguishing between placed and non-placed students.

In [298]:
print("Random Forest Train Accuracy:", rf_model.score(X_train, y_train))
print("Random Forest Test Accuracy:", rf_model.score(X_test, y_test))

Random Forest Train Accuracy: 0.8021539688871161
Random Forest Test Accuracy: 0.777511961722488


In [299]:
cv_scores_rf = cross_val_score(rf_model, X, y, cv=5)
print("Random Forest CV Accuracy:", cv_scores_rf.mean())

Random Forest CV Accuracy: 0.7635624103427691


In [300]:
rf_prob = rf_model.predict_proba(X_test)[:, 1]
auc_rf = roc_auc_score(y_test, rf_prob)
print("Random Forest ROC-AUC:", auc_rf)

Random Forest ROC-AUC: 0.8206191265892758


In [301]:
importance_rf = rf_model.feature_importances_

feature_df_rf = pd.DataFrame({
    "Feature": X.columns,
    "Importance": importance_rf
}).sort_values(by="Importance", ascending=False)

feature_df_rf

,Feature,Importance
4,AptitudeTestScore,0.244220
0,CGPA,0.177795
6,Technical_Skills,0.176033
2,Projects,0.156141
5,Communication_Skills,0.114879
3,Certifications,0.095888
1,Internship_Experience,0.027086
7,Internship_Flag,0.007958


The Random Forest model achieved a training accuracy of 80.21% and a testing accuracy of 77.75%. The small difference between these two values (approximately 2.5%) indicates that the model is not significantly overfitting and maintains good generalization performance on unseen data.
The 5-fold cross-validation accuracy of 76.36% is very close to the test accuracy.The ROC-AUC score of 0.8206 suggests that the model has strong discriminative ability in distinguishing between placed and non-placed students.

Final Model Selection – Random Forest

Random Forest was selected as the final model because it demonstrated better overall balance and generalization performance. It achieved a test accuracy of 77.75%, cross-validation accuracy of 76.36%, and a higher ROC-AUC score of 0.8206, outperforming CatBoost on overall stability.

Unlike CatBoost, Random Forest showed a smaller gap between training and testing accuracy, indicating lower overfitting and better reliability on unseen data. Therefore, Random Forest was chosen as the final model for placement prediction.

# Saving the Model

In [302]:
import joblib

joblib.dump(rf_model, "placement_random_forest_model.pkl")


['placement_random_forest_model.pkl']

In [303]:
joblib.dump(X_train.columns.tolist(), "placement_model_features.pkl")

['placement_model_features.pkl']

# Testing 

In [304]:
import joblib
import pandas as pd

rf_model = joblib.load("placement_random_forest_model.pkl")
features = joblib.load("placement_model_features.pkl")

In [305]:
sample = {
    "CGPA": 4.5,
    "Internship_Experience": 0,
    "Projects": 0,
    "Certifications": 0,
    "AptitudeTestScore": 30,
    "Communication_Skills": 1,
    "Technical_Skills": 1,
    "Internship_Flag": 0
}

df_input = pd.DataFrame([sample])
df_input = df_input[features]

prediction = rf_model.predict(df_input)
probability = rf_model.predict_proba(df_input)
print("Prediction:", prediction[0])
print("Placement Probability:", probability[0][1])

Prediction: 0
Placement Probability: 0.14562107541894542


In [306]:
sample = {
    "CGPA": 6.8,
    "Internship_Experience": 1,
    "Projects": 2,
    "Certifications": 1,
    "AptitudeTestScore": 65,
    "Communication_Skills": 6,
    "Technical_Skills": 6,
    "Internship_Flag": 1
}

df_input = pd.DataFrame([sample])
df_input = df_input[features]

prediction = rf_model.predict(df_input)
probability = rf_model.predict_proba(df_input)
print("Prediction:", prediction[0])
print("Placement Probability:", probability[0][1])

Prediction: 0
Placement Probability: 0.185195056286991


In [307]:
sample = {
    "CGPA": 8.9,
    "Internship_Experience": 2,
    "Projects": 4,
    "Certifications": 3,
    "AptitudeTestScore": 85,
    "Communication_Skills": 8,
    "Technical_Skills": 9,
    "Internship_Flag": 1
}

df_input = pd.DataFrame([sample])
df_input = df_input[features]

prediction = rf_model.predict(df_input)
probability = rf_model.predict_proba(df_input)
print("Prediction:", prediction[0])
print("Placement Probability:", probability[0][1])

Prediction: 1
Placement Probability: 0.8099188370136852


In [308]:
sample = {
    "CGPA": 9.2,
    "Internship_Experience": 0,
    "Projects": 1,
    "Certifications": 0,
    "AptitudeTestScore": 70,
    "Communication_Skills": 3,
    "Technical_Skills": 3,
    "Internship_Flag": 0
}

df_input = pd.DataFrame([sample])
df_input = df_input[features]

prediction = rf_model.predict(df_input)
probability = rf_model.predict_proba(df_input)
print("Prediction:", prediction[0])
print("Placement Probability:", probability[0][1])

Prediction: 0
Placement Probability: 0.17063660365230537


In [309]:
sample = {
    "CGPA": 5.8,
    "Internship_Experience": 2,
    "Projects": 5,
    "Certifications": 2,
    "AptitudeTestScore": 78,
    "Communication_Skills": 7,
    "Technical_Skills": 9,
    "Internship_Flag": 1
}
df_input = pd.DataFrame([sample])
df_input = df_input[features]

prediction = rf_model.predict(df_input)
probability = rf_model.predict_proba(df_input)
print("Prediction:", prediction[0])
print("Placement Probability:", probability[0][1])

Prediction: 1
Placement Probability: 0.5013937534596393


In [310]:
sample = {
    "CGPA": 7.2,
    "Internship_Experience": 1,
    "Projects": 2,
    "Certifications": 1,
    "AptitudeTestScore": 68,
    "Communication_Skills": 9,
    "Technical_Skills": 6,
    "Internship_Flag": 1
}

df_input = pd.DataFrame([sample])
df_input = df_input[features]

prediction = rf_model.predict(df_input)
probability = rf_model.predict_proba(df_input)
print("Prediction:", prediction[0])
print("Placement Probability:", probability[0][1])

Prediction: 0
Placement Probability: 0.1388877797011899


In [311]:
sample = {
    "CGPA": 6.5,
    "Internship_Experience": 3,
    "Projects": 2,
    "Certifications": 2,
    "AptitudeTestScore": 72,
    "Communication_Skills": 6,
    "Technical_Skills": 7,
    "Internship_Flag": 1
}

df_input = pd.DataFrame([sample])
df_input = df_input[features]

prediction = rf_model.predict(df_input)
probability = rf_model.predict_proba(df_input)
print("Prediction:", prediction[0])
print("Placement Probability:", probability[0][1])

Prediction: 0
Placement Probability: 0.3102452058027822


In [312]:
sample = {
    "CGPA": 6.0,
    "Internship_Experience": 0,
    "Projects": 1,
    "Certifications": 0,
    "AptitudeTestScore": 55,
    "Communication_Skills": 5,
    "Technical_Skills": 5,
    "Internship_Flag": 0
}

df_input = pd.DataFrame([sample])
df_input = df_input[features]

prediction = rf_model.predict(df_input)
probability = rf_model.predict_proba(df_input)
print("Prediction:", prediction[0])
print("Placement Probability:", probability[0][1])

Prediction: 0
Placement Probability: 0.18689853386689034


In [314]:
y.value_counts(normalize=True)

Placement_Label
0    0.641512
1    0.358488
Name: proportion, dtype: float64